src/util/invariant.h

```diff
@@ -81,25 +132,79 @@ class invariant_failedt: public std::logic_error
 
 
 #else
-// CPROVER_INVARIANT_PRINT_STACK_TRACE affects the implementation of
-// this function but not it's generation from the macro
 
-void check_invariant(
+void print_backtrace(
+  const char * const file,
+  const char * const function,
+  const int line,
+  std::ostream &out);
+
+std::string get_backtrace(
+  const char * const file,
+  const char * const function,
+  const int line);
+
+/// Takes a backtrace, gives it to the reason structure, then aborts, printing
+/// reason.what() (which therefore includes the backtrace).
+/// In future this may throw `reason` instead of aborting.
+/// \param file : C string giving the name of the file.
+/// \param function : C string giving the name of the function.
+/// \param line : The line number of the invariant
+/// \param reason : (derived from invariant_failedt) error report structure;
+///  its backtrace member will be set before it is used.
+template<class T>
+typename std::enable_if<std::is_base_of<invariant_failedt, T>::value>::type
```

Comment:
I took the true type because when we throw them we'll want to throw the actual type. If I throw an `invariant_failedt &` then I will attempt to copy-construct a the throw site (throws are by value). Such a copy constructor would discard information.